tools/tpm2_listpersistent.c

```diff
@@ -62,19 +92,37 @@ int readPublic(TSS2_SYS_CONTEXT *sapi_context,
     rval = Tss2_Sys_ReadPublic(sapi_context, objectHandle, 0, &outPublic, &name, &qualifiedName, &sessionsDataOut);
     if(rval != TPM_RC_SUCCESS)
     {
-        printf("\nTPM2_ReadPublic error: rval = 0x%0x\n\n",rval);
+        LOG_ERR("\nTPM2_ReadPublic error: rval = 0x%0x\n",rval);
         return -1;
     }
 
-    printf("  {\n");
-    printf("\tType: 0x%x\n ", outPublic.t.publicArea.type);
-    printf("\tHash algorithm(nameAlg): 0x%x\n ", outPublic.t.publicArea.nameAlg);
-    printf("\tAttributes: 0x%x\n", outPublic.t.publicArea.objectAttributes.val);
-    printf("  }\n");
+    TPMI_ALG_PUBLIC type = outPublic.t.publicArea.type;
+    TPMI_ALG_HASH nameAlg = outPublic.t.publicArea.nameAlg;
+    UINT32 attrs = outPublic.t.publicArea.objectAttributes.val;
+
+    if ((ctx.type && (type != ctx.type)) ||
+        (ctx.nameAlg && (nameAlg != ctx.nameAlg)))
```

Comment:
So TPM_ALG_ERROR aka 0 is that what you want here, ie if it's not zero run the filter. Perhaps initialize these to TPM_ALG_NULL? Not sure what/if the benefit is, i just want to make sure its clear that your checking against TPM_ALG_ERROR here.

Reply:
Yes, I noticed that TPM_ALG_ERROR is 0 but I thought it would be OK since either the members won't be initialized or the option would had been wrong and failed.

But now that you mention, probably I should init to TPM_ALG_NULL to check in the tpm2_tool_onstart() that the user didn't provide wrong arguments. And while looking at that I noticed a flaw in tpm2_alg_util_from_optarg().

If you provide a hash algorithm to the key option or viceversa, tpm2_alg_util_from_optarg() will happily convert the argument to a valid TPM_ALG value.

Reply:
@williamcroberts this is what I meant BTW:

```
$ tpm2_listpersistent 
persistent-handle[0]:0x81000000 key-alg:ecc hash-alg:sha256 object-attr:0x30072

$ tpm2_listpersistent -g foo
ERROR: Invalid hash algorithm, got "foo"

$ tpm2_listpersistent -g ecc
$
```

In any case this is independent to this change, so I'll just update whit your suggestion and we should fix tpm2_alg_util_from_optarg() as a separate issue.

Reply:
That's not an issue in ```tpm2_alg_util_from_optarg```, the caller should check if the algorithm is a hash algorithm via: ```tpm2_alg_util_is_hash_alg```

I thought about adding flags to the optarg conversion routine to specify what types of algorithms are allowed...

```
union flags {
  struct {
    UINT8 hash            : 1;
    UINT8 asymmetric : 1;
    UINT8 symmetric   : 1;
    UINT8 unused        : 5;
  };
UINT8 val;
};
```

And then based on the flags, check if its valid.

Reply:
@williamcroberts got it, I missed that function. I'll add a fixup then.